File: composeApp/src/androidMain/kotlin/com/humblesolutions/indsphinx/repository/BackendQuestionNotificationRepository.kt
package com.humblesolutions.indsphinx.repository

import android.util.Log
import com.google.firebase.Timestamp
import com.google.firebase.firestore.FieldValue
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.SetOptions
import com.humblesolutions.indsphinx.model.QuestionNotification
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.tasks.await

class BackendQuestionNotificationRepository : QuestionNotificationRepository {
    private val db = FirebaseFirestore.getInstance()

    override suspend fun fetchQuestionNotification(qnId: String): QuestionNotification? {
        val doc = db.collection("QuestionNotifications").document(qnId).get().await()
        if (!doc.exists()) return null
        val data = doc.data ?: return null
        val title = data["title"] as? String ?: return null
        val description = data["description"] as? String ?: ""
        @Suppress("UNCHECKED_CAST")
        val options = (data["options"] as? List<String>) ?: emptyList()
        val allowMultipleChoices = data["allowMultipleChoices"] as? Boolean ?: false
        val targetType = data["targetType"] as? String ?: ""
        @Suppress("UNCHECKED_CAST")
        val targetUserIds = (data["targetUserIds"] as? List<String>) ?: emptyList()
        val responseCount = (doc.getLong("responseCount") ?: 0L).toInt()
        val createdAt = (data["createdAt"] as? Timestamp)?.toDate()?.time ?: 0L
        val responseDeadline = (data["responseDeadline"] as? Timestamp)?.toDate()?.time
        return QuestionNotification(
            id = doc.id,
            title = title,
            description = description,
            options = options,
            allowMultipleChoices = allowMultipleChoices,
            targetType = targetType,
            targetUserIds = targetUserIds,
            responseCount = responseCount,
            createdAt = createdAt,
            responseDeadline = responseDeadline
        )
    }

    override fun observeResponseExists(qnId: String, uid: String): Flow<Boolean> = callbackFlow {
        val reg = db.collection("QuestionNotifications").document(qnId)
            .collection("responses").document(uid)
            .addSnapshotListener { snap, _ -> trySend(snap?.exists() == true) }
        awaitClose { reg.remove() }
    }

    override suspend fun submitResponse(
        qnId: String,
        uid: String,
        displayName: String,
        flatNo: String,
        recipientType: String,
        selectedOptions: List<String>,
        textResponse: String
    ) {
        val responseRef = db.collection("QuestionNotifications").document(qnId)
            .collection("responses").document(uid)

        val data = mapOf(
            "uid" to uid,
            "displayName" to displayName,
            "flatNo" to flatNo,
            "recipientType" to recipientType,
            "selectedOptions" to selectedOptions,
            "textResponse" to textResponse,
            "submittedAt" to FieldValue.serverTimestamp()
        )
        // The parent `responseCount` is maintained by the `onQuestionResponseCreated`
        // Cloud Function trigger — clients cannot write the parent doc because the
        // Firestore rule restricts QN parent writes to admins.
        Log.d(TAG, "submitResponse: qnId=$qnId uid=$uid optionsCount=${selectedOptions.size} hasText=${textResponse.isNotEmpty()}")
        try {
            responseRef.set(data, SetOptions.merge()).await()
            Log.d(TAG, "submitResponse: success qnId=$qnId uid=$uid")
        } catch (e: Exception) {
            Log.e(TAG, "submitResponse: FAILED qnId=$qnId uid=$uid", e)
            throw e
        }
    }

    private companion object {
        private const val TAG = "QuestionNotifFlow"
    }
}

File: iosApp/iosApp/BackendQuestionNotificationRepository.swift
import Foundation
import FirebaseFirestore

struct QuestionNotification: Identifiable {
    let id: String
    let title: String
    let description: String
    let options: [String]
    let allowMultipleChoices: Bool
    let targetType: String
    let targetUserIds: [String]
    let responseCount: Int
    let createdAt: Date
    let responseDeadline: Date?
}

class BackendQuestionNotificationRepository {
    private let db = Firestore.firestore()

    func fetchQuestionNotification(qnId: String) async throws -> QuestionNotification? {
        let doc = try await db.collection("QuestionNotifications").document(qnId).getDocument()
        guard doc.exists, let data = doc.data(), let title = data["title"] as? String else { return nil }
        let description = data["description"] as? String ?? ""
        let options = data["options"] as? [String] ?? []
        let allowMultipleChoices = data["allowMultipleChoices"] as? Bool ?? false
        let targetType = data["targetType"] as? String ?? ""
        let targetUserIds = data["targetUserIds"] as? [String] ?? []
        let responseCount = data["responseCount"] as? Int ?? 0
        let createdAt = (data["createdAt"] as? Timestamp)?.dateValue() ?? Date()
        let responseDeadline = (data["responseDeadline"] as? Timestamp)?.dateValue()
        return QuestionNotification(
            id: doc.documentID,
            title: title,
            description: description,
            options: options,
            allowMultipleChoices: allowMultipleChoices,
            targetType: targetType,
            targetUserIds: targetUserIds,
            responseCount: responseCount,
            createdAt: createdAt,
            responseDeadline: responseDeadline
        )
    }

    func observeResponseExists(
        qnId: String,
        uid: String,
        onChange: @escaping (Bool) -> Void
    ) -> ListenerRegistration {
        return db.collection("QuestionNotifications").document(qnId)
            .collection("responses").document(uid)
            .addSnapshotListener { snap, _ in onChange(snap?.exists == true) }
    }

    func submitResponse(
        qnId: String,
        uid: String,
        displayName: String,
        flatNo: String,
        recipientType: String,
        selectedOptions: [String],
        textResponse: String
    ) async throws {
        let responseRef = db.collection("QuestionNotifications").document(qnId)
            .collection("responses").document(uid)

        let data: [String: Any] = [
            "uid": uid,
            "displayName": displayName,
            "flatNo": flatNo,
            "recipientType": recipientType,
            "selectedOptions": selectedOptions,
            "textResponse": textResponse,
            "submittedAt": FieldValue.serverTimestamp()
        ]
        // The parent `responseCount` is incremented by the
        // `onQuestionResponseCreated` Cloud Function trigger — clients cannot
        // write the parent because the Firestore rule restricts QN parent
        // writes to admins.
        print("[QuestionNotifFlow] submitResponse: qnId=\(qnId) uid=\(uid) optionsCount=\(selectedOptions.count) hasText=\(!textResponse.isEmpty)")
        do {
            try await responseRef.setData(data, merge: true)
            print("[QuestionNotifFlow] submitResponse: success qnId=\(qnId) uid=\(uid)")
        } catch {
            print("[QuestionNotifFlow] submitResponse: FAILED qnId=\(qnId) uid=\(uid) error=\(error)")
            throw error
        }
    }
}
